go/pools/smartconnpool/pool.go

```diff
@@ -407,6 +432,23 @@ func (pool *ConnPool[C]) put(conn *Pooled[C]) {
 	}
 }
 
+// closeOnIdleLimitReached closes a connection if the number of idle connections (active - inuse) in the pool
+// exceeds the idleCount limit. It returns true if the connection is closed, false otherwise.
+func (pool *ConnPool[C]) closeOnIdleLimitReached(conn *Pooled[C]) bool {
```

Comment:
Would it be better to not always immediately close, or just close with some random chance, so the connection churn is not really aggressive when there's high active transaction churn?

Reply:
We have to maintain a pool of some free available connections, otherwise there will be time spend in acquiring the connection for executing the query. 
Even during high transactions, there will be log of `get` and `put` call that will happen on the pool, so we have to wait till the free connections become more than a certain limit before we start closing them.
Otherwise the application will see degradation in performance on high QPS

Reply:
Right, I said "not always". By which I meant, randomize closing after the idle time had reached.

Reply:
Another way to look at it would be to make the idle time random, say between 1x and 2x of the config value, so we don't try and close lots of connections at the same time, after a spike in load.

Reply:
We already have support for idle time. This is implementation for idle count. You can choose to use either or both settings on the pool.

Reply:
Yes, but the idle time and counts are hard thresholds, which will cause large numbers of connections to get closed at the same time, if a large number were opened at the same time, which is not necessary, and could be smoothed out a little.

Reply:
It is not like maximum connection lifetime which we already smoothed out.
In case of idle connection, it is reasonable to churn them as and when they exceed the limit. 
There is no active user looking for those connections, therefore they became idle.